bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigotWorldManager.java

```diff
@@ -111,14 +127,16 @@ public static int getLegacyBlock(GeyserSession session, int x, int y, int z, boo
         if (isViaVersion) {
             Block block = Bukkit.getPlayer(session.getPlayerEntity().getUsername()).getWorld().getBlockAt(x, y, z);
             // Black magic that gets the old block state ID
-            int oldBlockId = (block.getType().getId() << 4) | (block.getData() & 0xF);
-            // Convert block state from old version -> 1.13 -> 1.13.1 -> 1.14 -> 1.15 -> 1.16 -> 1.16.2
-            int thirteenBlockId = us.myles.ViaVersion.protocols.protocol1_13to1_12_2.data.MappingData.blockMappings.getNewId(oldBlockId);
-            int thirteenPointOneBlockId = Protocol1_13_1To1_13.getNewBlockStateId(thirteenBlockId);
-            int fourteenBlockId = us.myles.ViaVersion.protocols.protocol1_14to1_13_2.data.MappingData.blockStateMappings.getNewId(thirteenPointOneBlockId);
-            int fifteenBlockId = us.myles.ViaVersion.protocols.protocol1_15to1_14_4.data.MappingData.blockStateMappings.getNewId(fourteenBlockId);
-            int sixteenBlockId = us.myles.ViaVersion.protocols.protocol1_16to1_15_2.data.MappingData.blockStateMappings.getNewId(fifteenBlockId);
-            return MappingData.blockStateMappings.getNewId(sixteenBlockId);
+            int blockId = (block.getType().getId() << 4) | (block.getData() & 0xF);
+            // Convert block state from old version (1.12.2) -> 1.13 -> 1.13.1 -> 1.14 -> 1.15 -> 1.16 -> 1.16.2
+            blockId = ProtocolRegistry.getProtocol(Protocol1_13To1_12_2.class).getMappingData().getNewBlockId(blockId);
+            List<Pair<Integer, Protocol>> protocolList = ProtocolRegistry.getProtocolPath(CLIENT_PROTOCOL_VERSION,
+                    ProtocolVersion.v1_13.getId());
+            for (int i = protocolList.size() - 1; i >= 0; i--) {
+                if (!protocolList.get(i).getValue().hasMappingDataToLoad()) continue;
```

Comment:
The method *may* be overridden in other cases, so relying on this is more future-proof :p
```suggestion
                if (protocolList.get(i).getValue().getMappingData() == null) continue;
```